dspy/adapters/base.py

```diff
@@ -42,55 +44,271 @@ def __call__(
-                )
-
             if output_logprobs is not None:
                 value["logprobs"] = output_logprobs
 
             values.append(value)
 
         return values
 
-    @abstractmethod
     def format(
         self,
         signature: Type[Signature],
         demos: list[dict[str, Any]],
         inputs: dict[str, Any],
     ) -> list[dict[str, Any]]:
-        raise NotImplementedError
+        """Format the input messages for the LM call.
 
-    @abstractmethod
-    def parse(self, signature: Type[Signature], completion: str) -> dict[str, Any]:
-        raise NotImplementedError
+        This method converts the DSPy structured input along with few-shot examples and conversation history into
+        multiturn messages as expected by the LM. For custom adapters, this method can be overridden to customize
+        the formatting of the input messages.
+
+        In general we recommend the messages to have the following structure:
+        ```
+        [
+            {"role": "system", "content": system_message},
+            # Begin few-shot examples
+            {"role": "user", "content": few_shot_example_1_input},
+            {"role": "assistant", "content": few_shot_example_1_output},
+            {"role": "user", "content": few_shot_example_2_input},
+            {"role": "assistant", "content": few_shot_example_2_output},
+            ...
+            # End few-shot examples
+            # Begin conversation history
+            {"role": "user", "content": conversation_history_1_input},
+            {"role": "assistant", "content": conversation_history_1_output},
+            {"role": "user", "content": conversation_history_2_input},
+            {"role": "assistant", "content": conversation_history_2_output},
+            ...
+            # End conversation history
+            {"role": "user", "content": current_input},
+        ]
+
+        And system message should contain the field description, field structure, and task description.
+        ```
+
+
+        Args:
+            signature: The DSPy signature for which to format the input messages.
+            demos: A list of few-shot examples.
+            inputs: The input arguments to the DSPy module.
+
+        Returns:
+            A list of multiturn messages as expected by the LM.
+        """
+        inputs_copy = dict(inputs)
+        messages = []
+        system_message = (
+            f"{self.format_field_description(signature)}\n"
+            f"{self.format_field_structure(signature)}\n"
+            f"{self.format_task_description(signature)}"
+        )
+        messages.append({"role": "system", "content": system_message})
+        messages.extend(self.format_demos(signature, demos))
+        conversation_history = self.format_conversation_history(signature, inputs_copy)
+        if conversation_history:
+            # Conversation history and current input
+            messages.extend(conversation_history)
```

Comment:
How do we handle cases where the *demos* have multi-turn history?

Reply:
it will be a single turn entity with history as one field, for example:

```
User message:

This is an example of the task, though some input or output fields are not supplied.

[[ ## question ## ]]
What is the capital of France?

[[ ## history ## ]]
[{"question": "What is the capital of Germany?", "answer": "Berlin"}]

Respond with the corresponding output fields, starting with the field `[[ ## reasoning ## ]]`, then `[[ ## answer ## ]]`, and then ending with the marker for `[[ ## completed ## ]]`.


Assistant message:

[[ ## reasoning ## ]]
Not supplied for this particular example. 

[[ ## answer ## ]]
Paris

```